pkg/instance/service.go

```diff
@@ -76,43 +80,267 @@ func (s service) SaveInstance(instance *model.DeploymentInstance) error {
+			return nil, err
+		}
+		for _, requiredStack := range stack.Requires {
+			err := g.AddEdge(src.StackName, requiredStack.Name)
+			if err != nil {
+				if errors.Is(err, graph.ErrEdgeAlreadyExists) {
+					return nil, fmt.Errorf("instance %q requires %q more than once", src.Name, requiredStack.Name)
+				} else if errors.Is(err, graph.ErrEdgeCreatesCycle) {
+					return nil, fmt.Errorf("edge from instance %q to instance %q creates a cycle", src.Name, requiredStack.Name)
+				}
+				return nil, fmt.Errorf("failed adding edge from instance %q to instance %q: %v", src.Name, requiredStack.Name, err)
+			}
+		}
+	}
+
+	return g, nil
+}
+
+func (s service) resolveParameters(deployment *model.Deployment) error {
+	for _, instance := range deployment.Instances {
+		instanceParameters := instance.Parameters
+
+		stack, err := s.stackService.Find(instance.StackName)
+		if err != nil {
+			return err
+		}
+
+		err = rejectConsumedParameters(instanceParameters, stack)
+		if err != nil {
+			return err
+		}
+
+		err = rejectNonExistingParameters(instanceParameters, stack)
+		if err != nil {
+			return err
+		}
+
+		addUnsuppliedParameters(stack, instanceParameters)
+
+		for name, parameter := range instanceParameters {
+			if !stack.Parameters[name].Consumed {
+				continue
+			}
+
+			for _, requiredStack := range stack.Requires {
+				// consume from instance parameters
+				sourceInstance := deployment.FindInstanceByStackName(requiredStack.Name)
+				if sourceInstance == nil {
+					return errdef.NewNotFound("failed to find required instance %q of instance %q", sourceInstance.Name, instance.Name)
+				}
+
+				if sourceInstanceParameter, ok := sourceInstance.Parameters[name]; ok {
+					parameter.Value = sourceInstanceParameter.Value
+				}
+
+				// consume from provider
+				if provider, ok := requiredStack.ParameterProviders[name]; ok {
+					value, err := provider.Provide(*sourceInstance)
+					if err != nil {
+						return fmt.Errorf("failed to provide value for instance %q parameter %q: %v", instance.Name, name, err)
+					}
+					parameter.Value = value
+				}
+
+				instanceParameters[name] = parameter
+			}
+		}
+	}
+
+	return nil
+}
+
+func rejectConsumedParameters(instanceParameters model.DeploymentInstanceParameters, stack *model.Stack) error {
+	var errs []error
+	for name := range instanceParameters {
+		if stack.Parameters[name].Consumed {
+			errs = append(errs, fmt.Errorf("consumed parameters can't be supplied by the user: %s", name))
+		}
+	}
+	return errors.Join(errs...)
+}
+
+func rejectNonExistingParameters(instanceParameters model.DeploymentInstanceParameters, stack *model.Stack) error {
+	var errs []error
+	for name := range instanceParameters {
+		if _, ok := stack.Parameters[name]; !ok {
+			errs = append(errs, fmt.Errorf("parameter not found on stack: %s", name))
+		}
+	}
+	return errors.Join(errs...)
+}
+
+func addUnsuppliedParameters(stack *model.Stack, instanceParameters model.DeploymentInstanceParameters) {
```

Comment:
I would make this only about default values for `!Consumed` parameters => `addDefaultParameterValues`.

I find the interplay between `addUnsuppliedParameters` and resolving consumed parameters above odd/confusing 🤔 By that I mean that `instanceParameters[name] = instanceParameter`.